rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java

```diff
@@ -1585,10 +1585,10 @@ private J.VariableDeclarations visitVariables(List<VariableTree> nodes, Space fm
 
             Space namedVarPrefix = sourceBefore(n.getName().toString());
 
-            JavaType type = typeMapping.type(n);
+            JavaType.Variable type = typeMapping.variableType(n.sym);
             J.Identifier name = new J.Identifier(randomId(), EMPTY, Markers.EMPTY, emptyList(), n.getName().toString(),
-                    type instanceof JavaType.Variable ? ((JavaType.Variable) type).getType() : type,
-                    type instanceof JavaType.Variable ? (JavaType.Variable) type : null);
+                    type != null ? type.getType() : null,
+                    type);
```

Comment:
Should we do the same for our Java 21 parser to carry this improvement forward?
https://github.com/openrewrite/rewrite/blob/628e0b4b50831a761119eadc30b7748e9d60333c/rewrite-java-21/src/main/java/org/openrewrite/java/isolated/ReloadableJava21ParserVisitor.java#L1586-L1588

Reply:
Ah we posted the same insight in the same minute. :)